File: test-suite-python/src/test/python/micronaut/flyway/docs/CustomFlywayTypesFactory.py
from micronaut.context.annotation import Requires

# tag::imports[]
from jakarta.inject import Named, Singleton
from micronaut.context.annotation import Factory
from org.flywaydb.core.api.migration import BaseJavaMigration, Context, JavaMigration
# end::imports[]


@Requires(property="spec.name", pattern="CustomFlywayTypesFactoryTest|BooksFlywayConfigurationCustomizerTest")
# tag::clazz[]
@Factory
class CustomFlywayTypesFactory:

    @Named("books")  # <1>
    @Singleton
    def books_migrations(self) -> list[JavaMigration]:
        return [V3__Migrate_books()]  # <2>


class V3__Migrate_books(BaseJavaMigration):  # <3>

    def migrate(self, context: Context) -> None:
        pass  # Execute migration
# end::clazz[]

File: test-suite-java/src/test/java/io/micronaut/flyway/docs/CustomFlywayTypesFactory.java
package io.micronaut.flyway.docs;

import io.micronaut.context.annotation.Requires;

// tag::imports[]
import io.micronaut.context.annotation.Factory;
import jakarta.inject.Named;
import jakarta.inject.Singleton;
import org.flywaydb.core.api.migration.BaseJavaMigration;
import org.flywaydb.core.api.migration.Context;
import org.flywaydb.core.api.migration.JavaMigration;
// end::imports[]

@Requires(property = "spec.name", pattern = "CustomFlywayTypesFactoryTest|BooksFlywayConfigurationCustomizerTest")
// tag::clazz[]
@Factory
public class CustomFlywayTypesFactory {

    @Named("books") // <1>
    @Singleton
    public JavaMigration[] booksMigrations() {
        return new JavaMigration[] { new V3__Migrate_books() }; // <2>
    }

    static class V3__Migrate_books extends BaseJavaMigration { // <3>

        @Override
        public void migrate(Context context) throws Exception {
            // Execute migration
        }
    }
}
// end::clazz[]
